File: backend/src/main/java/fr/iut/valence/but/info/ctvr/backend/service/BusService.java
package fr.iut.valence.but.info.ctvr.backend.service;

import fr.iut.valence.but.info.ctvr.backend.dto.BusIdentity;
import fr.iut.valence.but.info.ctvr.backend.model.Bus;
import fr.iut.valence.but.info.ctvr.backend.repository.BusRepository;
import lombok.AllArgsConstructor;
import org.apache.coyote.BadRequestException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@AllArgsConstructor
public class BusService {
    private final BusRepository repository;

    public Iterable<Bus> getList() {
        return this.repository.findAllByOrderByNumber();
    }

    public void create(Bus bus) {
        this.repository.save(bus);
    }

    @Transactional
    public void update(Bus bus, String registration) throws BadRequestException {
        int rowCount = this.repository.update(bus, registration);
        if (rowCount != 1) {
            throw new BadRequestException("Bus introuvable");
        }
        this.repository.save(bus);
    }

    @Transactional
    public void delete(String registration) throws BadRequestException {
        int count = this.repository.deleteByRegistration(registration);
        if (count != 1) {
            throw new BadRequestException("Bus introuvable");
        }
    }

    public List<BusIdentity> getIdentities() {
        return this.repository.getIdentities();
    }

    public BusIdentity getIdentity(String registration) {
        return this.repository.getIdentity(registration);
    }
}

File: frontend/src/app/pages/bus-page/bus.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Bus, BusIdentity } from './bus';

@Injectable({
    providedIn: 'root',
})
export class BusService {
    constructor(private http: HttpClient) {}

    public getList() {
        return this.http.get<Bus[]>('/api/bus');
    }

    public create(bus: Bus) {
        return this.http.post('/api/bus', bus, { responseType: 'text' });
    }

    public update(bus: Bus, registration: string) {
        return this.http.put(`/api/bus/${registration}`, bus, {
            responseType: 'text',
        });
    }

    public delete(registration: string) {
        return this.http.delete(
            `/api/bus/${encodeURIComponent(registration)}`,
            { responseType: 'text' }
        );
    }

    public getIdentities() {
        return this.http.get<BusIdentity[]>('/api/bus/identities');
    }

    public getIdentity(registration: string) {
        return this.http.get<BusIdentity>(
            `/api/bus/${encodeURIComponent(registration)}/identity`
        );
    }
}
